From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Thu, 19 Feb 2026 04:43:05 +0100
Subject: document-portal: Convert early from app-id to permissions-id

In this way we avoid doing it at the end of the chain in the permission
portal

Origin: https://github.com/flatpak/xdg-desktop-portal/pull/1764
---
 document-portal/document-portal.c | 16 ++++++++--------
 document-portal/document-store.c  |  4 +++-
 2 files changed, 11 insertions(+), 9 deletions(-)

diff --git a/document-portal/document-portal.c b/document-portal/document-portal.c
index b702b1e..ed2dcab 100644
--- a/document-portal/document-portal.c
+++ b/document-portal/document-portal.c
@@ -137,7 +137,7 @@ portal_grant_permissions (GDBusMethodInvocation *invocation,
                           XdpAppInfo            *app_info)
 {
   const char *target_app_id;
-  const char *target_permissions_id;
+  g_autofree char *target_permissions_id = NULL;
   const char *id;
   g_autofree const char **permissions = NULL;
   DocumentPermissionFlags perms;
@@ -159,7 +159,7 @@ portal_grant_permissions (GDBusMethodInvocation *invocation,
         return;
       }
 
-    target_permissions_id = target_app_id;
+    target_permissions_id = permissions_id_for_app_id (target_app_id);
     if (!xdp_is_valid_permissions_id (target_permissions_id))
       {
         g_dbus_method_invocation_return_error (invocation,
@@ -202,7 +202,7 @@ portal_revoke_permissions (GDBusMethodInvocation *invocation,
 {
   const char *permissions_id = xdp_app_info_get_permissions_id (app_info);
   const char *target_app_id;
-  const char *target_permissions_id;
+  g_autofree char *target_permissions_id = NULL;
   const char *id;
   g_autofree const char **permissions = NULL;
   GError *error = NULL;
@@ -224,7 +224,7 @@ portal_revoke_permissions (GDBusMethodInvocation *invocation,
         return;
       }
 
-    target_permissions_id = target_app_id;
+    target_permissions_id = permissions_id_for_app_id (target_app_id);
     if (!xdp_is_valid_permissions_id (target_permissions_id))
       {
         g_dbus_method_invocation_return_error (invocation,
@@ -699,7 +699,7 @@ portal_add_full (GDBusMethodInvocation *invocation,
   g_autoptr(GVariant) array = NULL;
   guint32 flags;
   const char *target_app_id;
-  const char *target_permissions_id;
+  g_autofree char *target_permissions_id = NULL;
   g_autofree const char **permissions = NULL;
   DocumentPermissionFlags target_perms;
   gsize n_args;
@@ -759,7 +759,7 @@ portal_add_full (GDBusMethodInvocation *invocation,
         fd[i] = -1;
     }
 
-  target_permissions_id = target_app_id;
+  target_permissions_id = permissions_id_for_app_id (target_app_id);
   ids = document_add_full (fd, NULL, NULL, documents_flags, n_args, app_info,
                            target_permissions_id, target_perms, &error);
 
@@ -978,7 +978,7 @@ portal_add_named_full (GDBusMethodInvocation *invocation,
   guint32 flags = 0;
   const char *filename;
   const char *target_app_id;
-  const char *target_permissions_id;
+  g_autofree char *target_permissions_id = NULL;
   g_autofree const char **permissions = NULL;
   g_autofree char *id = NULL;
   g_autofree char *path = NULL;
@@ -1072,7 +1072,7 @@ portal_add_named_full (GDBusMethodInvocation *invocation,
 
     XDP_AUTOLOCK (db);
 
-    target_permissions_id = target_app_id;
+    target_permissions_id = permissions_id_for_app_id (target_app_id);
 
     if (as_needed_by_app &&
         app_has_file_access (target_permissions_id, target_perms, path))
diff --git a/document-portal/document-store.c b/document-portal/document-store.c
index 030c30c..441357f 100644
--- a/document-portal/document-store.c
+++ b/document-portal/document-store.c
@@ -85,8 +85,10 @@ document_entry_has_permissions_by_app_id (PermissionDbEntry       *entry,
                                           DocumentPermissionFlags  perms)
 {
   DocumentPermissionFlags current_perms;
+  g_autofree char *permissions_id = NULL;
 
-  current_perms = document_entry_get_permissions_by_app_permissions_id (entry, app_id);
+  permissions_id = permissions_id_for_app_id (app_id);
+  current_perms = document_entry_get_permissions_by_app_permissions_id (entry, permissions_id);
 
   return (current_perms & perms) == perms;
 }
